core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java

```diff
@@ -449,14 +450,31 @@ public Path getWorkingDirectory() {
    * Sets up a lazy connection to Alluxio through mFileSystem. This must be called before client
    * operations.
    *
-   * If it is called twice on the same object an exception will be thrown.
+   * If it is called twice on the same object concurrently, one of the thread will do the
+   * initialization work, the other will wait until initialization is done.
+   * If it is called after initialized, then this is a noop.
    */
   @Override
   public void initialize(URI uri, org.apache.hadoop.conf.Configuration conf) throws IOException {
     Preconditions.checkArgument(uri.getScheme().equals(getScheme()),
         PreconditionMessage.URI_SCHEME_MISMATCH.toString(), uri.getScheme(), getScheme());
-    Preconditions.checkArgument(mInitialized.compareAndSet(false, true), "Cannot invoke "
-        + "initialize() more than once");
+
+    if (mFileSystem != null) {
+      return;
+    }
+
+    if (mInitializationLock.writeLock().tryLock()) {
```

Comment:
can we just simplify this and put this initialization in a `synchronized` block of an object?

Reply:
considering that there shouldn't be lots of contentions for initializing, yes.